### PR TITLE
Algorithmic improvements

### PR DESCRIPTION
**Pull request by joekarl**
Just some things to think about (and some poorly written example code)

The original implementation of deePool methods have an O(n) time
complexity (in particular use/recycle).
This change makes those methods O(1) and vastly simplifies the code
It makes the concession of sacrificing ordering while growing the pool
and recycling objects (this is OK as the deePool didn't specify that
there was supposed to be any ordering semantics).

Tests are still busted (due to them relying on un-specified ordering
semantics) but functionally all works.
Also forgive the tabs/spaces mismatch. 

Here's the difference in perf numbers:
Original implementation:
```
deePool Performance Benchmarks
deePool.create()... complete: 6,644,319 ops/sec (391771 samples)
use() + recycle(): small pool (10)... complete: 9,704,383 ops/sec (533343 samples)
use() + recycle(): medium pool (10,000)... complete: 14,060 ops/sec (726 samples)
use() + recycle(): large pool (1,000,000)... complete: 132 ops/sec (7 samples)
use(): growing pool... complete: 2,662,634 ops/sec (397408 samples)
grow(): small increments... complete: 4,825 ops/sec (5000 samples)
grow(): medium increments... complete: 1,395 ops/sec (150 samples)
grow(): large increments... complete: 92 ops/sec (90 samples)
grow(): exponential (doubling)... complete: 49 ops/sec (19 samples)
use() + recycle(): interleaved... complete: 1,851,973 ops/sec (203786 samples)
use() + recycle(): multiple of each... <doesn't finish in reasonable time...>
```

0(1) implementation:
```
deePool Performance Benchmarks
deePool.create()... complete: 7,929,288 ops/sec (414483 samples)
use() + recycle(): small pool... complete: 44,645,133 ops/sec (2362781 samples)
use() + recycle(): medium pool... complete: 41,821,068 ops/sec (2356688 samples)
use() + recycle(): large pool... complete: 41,899,579 ops/sec (2416356 samples)
use(): growing pool... complete: 2,555,250 ops/sec (238327 samples)
grow(): small increments... complete: 708,790 ops/sec (162082 samples)
grow(): medium increments... complete: 7,729 ops/sec (2504 samples)
grow(): large increments... complete: 367 ops/sec (90 samples)
grow(): exponential (doubling)... complete: 65 ops/sec (19 samples)
use() + recycle(): interleaved... complete: 2,192,863 ops/sec (245557 samples)
use() + recycle(): multiple of each... complete: 35,301 ops/sec (5000 samples)
```

Note the perf decrease on `use() + recycle()` as the pool gets larger in the original implementation.
The ops/sec drop is almost exactly linear with respect the increase in pool size (indicating O(n))

### CODE REVIEW

**Comment by joekarl** [comment]
Continuing conversation from twitter:
The tl;dr is this: 
- yes, `use()` is designed to be O(1) for best case, but worst case is still O(n): `use()` will attempt to traverse all elements in the pool before it can figure out that it needs to grow the pool
- `recycle()` is O(n) when `nextOpenSlot` is null (when the pool is full or when it hasn't had any objects recycled back to it or hasn't been grown with objects in use)  

This can be visualized by adding in a simple log statement to count iterations of the 'searching for free/empty slots' here https://github.com/getify/deePool/blob/master/lib/deePool.src.js#L106 and here https://github.com/getify/deePool/blob/master/lib/deePool.src.js#L57

This can also be seen in the perf numbers themselves, as the pool gets larger, the number of iterations the benchmark can run gets smaller (ie it's bound by the size of the pool).

**Comment by joekarl** [comment]
As a heads up, the test failures in this PR are due to tests making assumptions about ordering of recycled objects. The O(1) implementation in this PR has different ordering semantics (due to swapping pointers around the array backed pool).

**Comment by getify** [comment]
IIUC, this algorithm means that:

```js
var o = pool.use();
pool.recycle(o);
var p = pool.use();

o === p;   // true
```

Is that correct? This is basically a stack implementation, right?

--------

I started originally with such an algorithm, but decided not to pursue it, after working on this a bit. The reason was related to the idea that it seems like an unhealthy pool if all the churn is in one spot and other elements in the pool are potentially never being used.

By unhealthy, what I mean is that the overall performance of the app may suffer. If `use()` and `recycle()` are being interleaved (the expected normal case), `use(..)` will most of the time return an object that was just recycled, instead of (in general) returning an object that may not have ever been used yet.

Consider the cost (not at all being factored in the benchmarks) of needing to condition/reset an object. Let's say you reset every object before you reinsert. Great, you pay that cost every object, every time. Predictable. Won't differ at all between your implementation and mine.

But what if, knowing that the pool works in round-robin instead of stack, you know that for the first n times you'll pull from the pool, you're going to get an unused object each time. There, instead of reseting on recycle, you check an object after `use()` returns it, and you only reset it if it needs to be reset. With that strategy, you'll be able to avoid the cost of reset for the first n usages of the pool.

IOW, in the best case, my implementation could eek out more overall performance.

I'm not saying I don't want to go in the direction of your implementation, I'm just describing why I considered it and didn't, originally.

**Comment by getify** [comment]
Another thing to point out... one intention of the original algorithm was that the pool did not hold onto a reference for an object once it had been handed out... this is because, in general, it should be the freedom of the user to `use()` an object, then throw away the reference (instead of recycling it), letting it then by GC'd if they wanted. Or, instead of resetting an object, you might decide it's best to just recreate it before recycling it.

In your implementation, a reference to the original object from the pool is kept in the pool, meaning that it's possible (depending on usage pattern) that you've recreated or thrown away an object but it didn't get GC'd the way you expect.

I would prefer to have used a WeakMap, but since the performance of those is still terrible, I did the next best thing and made sure that the pool doesn't hold onto references once it hands them out.

**Comment by joekarl** [comment]
So I hear you on fearing the cost of resetting the objects as the travel in/out of the pool. The issue is if you're doing a round robin, assuming sufficient usage you're going to hit all of the objects in the pool anyways and end up with the same characteristic of having to reset objects all the time. And that's at the cost of having to round robin through all of your objects which means traversing the backing array (or at least part of it). 

As for reset cost, I'd bet in practice that most objects in the position of having to be pooled rarely need to be reset before used again. Take a vector or some sort of mathy object. You'll generally set all of it's member variables anyways so no reset is required. As for other objects (say game objects) these will last over the course of many frames of gameplay so the reset cost is amortized across those frames (so hardly becomes a concern).

**Comment by joekarl** [comment]
Ok, so the pool holding onto a reference or not is an entirely different matter, Based on the README this wasn't a feature of the lib so I considered that it was fair game.

**Comment by getify** [comment]
> Based on the README this wasn't a feature of the lib

Yeah, as well with the ordering, these things aren't mentioned in the README because this lib is brand new and I'm still figuring out what it should and shouldn't do. I'm describing here what original intent is, for the purposes of discussion, but I hadn't committed to any of it yet. :)

**Comment by joekarl** [comment]
>> Based on the README this wasn't a feature of the lib
>
>Yeah, as well with the ordering, these things aren't mentioned in the README because this lib is brand new and I'm still figuring out what it should and shouldn't do. I'm describing here what original intent is, for the purposes of discussion, but I hadn't committed to any of it yet. :)

Heh, yeah. Well sounds like you've got a plan :) Just wanted to throw out there that as currently setup this is O(n) and could easily be implemented as O(1). For a an interesting parallel, you're getting close to the idea of a [Free List](https://en.wikipedia.org/wiki/Free_list) (which I'm not sure makes tons of sense to do in JS because of the whole non-contiguous object memory thing).

**Comment by getify** [comment]
It's quite possible my instinct is wrong here on the idea of round-robin being better than stack. It certainly feels less "healthy" (for some definition of that word) to almost starve the pool by use/recycle staying localized to one part of the pool. But other than the potential recondition costs, I don't have any other hard analysis to back that up.

What I am pretty sure of is, the other implementations I found in other libs, they literally used the array as a stack, and did `push()` / `pop()` on it, and that kind of churn of the array is overhead I do not want the pool to pay. But a virtualized stack with a pointer is at least reasonable to consider. :)

**Comment by joekarl** [comment]
Yeah, so the I gather the whole point of the pool is to trade space for time (in this case memory for GC pauses). So I'm not sure you need to be concerned about having unused objects lying around. You just need enough for any point in the game (and memory is cheap for most games, even in a browser). 

And yeah, like you noted, push/pop on a real array is gonna be a bad time (well maybe... I bet the vm does some optimizations of array sizes and whatnot internally...). This is basically what one would end up doing in C or something where you're keeping track of a bunch of pointers on the heap. Is a pretty

**Comment by joekarl** [comment]
Anyways, just thoughts, I'll let you carry on :)

**Comment by joekarl** [comment]
Oh, one last thing, `recycle()` is definitely doing O(n) for normal behavior (it's really the thing contributing to the overall O(n) behavior in the perf benchmarks). So probably need to look at that at least.

**Comment by getify** [comment]
As for the "worst case" of O(n) in the current implementation... I think the below analysis is accurate, but double-check me.

The full scan only happens for `use(..)` when you pull out the last free object in the pool (where the next time you call `use()` it's going to `grow()` and double the size), and then the scan only happens once. IOW, that scan only happens when the pool grows because you exhaust the pool entirely via `use()`. There are other times when `use(..)` is O(>1), but it's still O(1) even if it's O(4).

For `recycle()`, the O(n) scan happens (once) when you recycle() the last used object back into the pool. Again, there are other O(>1) scans, but these are all O(1). In the case where you, unadvisedly, decide to `recycle()` past the size of the pool (basically, adding extra objects in) and grow the pool by 1 each time, you also get the full O(n) scan each time.

I think in all other cases, `use()` and `recycle()` behave in O(1) time, right?

**Comment by joekarl** [comment]
Yeah, that was what I came away with as well for `use` and `recycle`.
The `use` O(n) is actually more insidious because you have to O(n) there, then O(n) again in the `grow` call (the latter can't be avoided because you can only allocate a single object at a time).

I wonder if you could remove the worst case in `recycle` if you walked backwards through the array instead of forwards...

File: lib/deePool.src.js [hunk 1]
```diff
@@ -29,25 +29,25 @@
...
-      }
-      return freePool[freePool.length - (freePoolEnd--)];
+			if (freePoolEnd == -1) {
+				if (freePool.length == 0) {
+					grow(5);
+				} else {
+					grow();
+				}
+			}
+			return freePool[freePool.length - 1 - (freePoolEnd--)];
```
Comment (getify) [inline lib/deePool.src.js]: do we need to subtract `freePoolEnd` from the length, or could it just be the pointer to the location itself... like `return freePool[ freePoolEnd-- ];`

File: lib/deePool.src.js [hunk 2]
```diff
@@ -29,25 +29,25 @@
...
 		}
 
 		function recycle(obj) {
-      freePool[freePool.length - (++freePoolEnd)] = obj;
+			freePool[freePool.length - (++freePoolEnd)] = obj;
```
Comment (getify) [inline lib/deePool.src.js]: ditto question here as comment on line 39.

**Comment by getify** [comment]
> it's really the thing contributing to the overall O(n) behavior in the perf benchmarks

specifically, the fact that the perf benchmarks do a `use()` followed by a `recycle()`, that's what's triggering the full O(n) scan every time. These benchmarks are currently pretty flawed in the sense that they don't really simulate well the kind of real behavior they'd get in actual applications.

File: lib/deePool.src.js (hunk 1 above)
Comment (joekarl) [inline lib/deePool.src.js]: Yeah, so that's just so when we `grow` the pool we can push onto the end of the backing array instead of unshifting to the front of it. Testing showed major slowdowns with growing the array from the front instead of the end (which makes sense because the vm probably can't take advantage of memcopy efficiently under the hood).

File: lib/deePool.src.js (hunk 2 above)
Comment (joekarl) [inline lib/deePool.src.js]: (see previous comment)

File: lib/deePool.src.js (hunk 1 above)
Comment (getify) [inline lib/deePool.src.js]: Couldn't freePoolEnd be renamed freePoolIndex, start out at 0, increase on use(), decrease on recycle(), and just be checked against freePool.length instead of -1 to know when we need to grow?

**Comment by getify** [comment]
inspired by your simpler algorithm ideas, i just rewrote the code. thoughts welcome. :)
